Auto Tag Confirmation

Flashy78 opened the pull request:
Resolves #3909 

Implements most of the suggestions from the ticket, except for a button to hide the dialogue in the future. I didn't see any similar functionality in Stash, and don't want to incorrectly build new features.

- Updated descriptions for Auto Tag and Identify with more explanation
- Auto Tag buttons should all have the confirmation dialog
- For new users, in the Settings their Performers and Studios will be disabled by default for Auto Tag
- For new users, removed Auto Tag as a default source from Identify

Other

- Refactored `DirectorySettingDialog.tsx` to allow customization and potentially hide the directory picker. This was copied from how the Clean task dialogs were doing their own directory picking, and the Clean task was refactored to use the updated `DirectorySettingDialog.tsx`. So now you can have a directory picking dialog that doesn't show you any directories, but all of the directory picking code is no longer duplicated for the Clean task.
- Added slight padding to the bottom of the directory listing to remove an inline scroll bar that was showing on my system

DingDongSoLong4 commented:
Very nice, this will definitely help prevent accidental presses or unaware users from activating these irreversible features.

I have just two comments:
- It would be nice if the Auto Tag dialogs on the studios, performers and tags pages showed different messages, explaining that it will Auto Tag for that specific studio/performer/tag only.
- Since there's no big warning icon, I think adding a simple "Are you sure you want to continue?" message to the Auto Tag confirmation dialogs would help to emphasize that the action is irreversible.

WithoutPants commented on ui/v2.5/src/components/Settings/Tasks/DataManagementTasks.tsx:
```diff
@@ -232,6 +133,45 @@ export const DataManagementTasks: React.FC<IDataManagementTasks> = ({
+        msg = intl.formatMessage({ id: "actions.tasks.clean_confirm_message" });
+      }
+
+      return (
+        <DirectorySelectionDialog
+          allowPathSelection={dialogOpen.clean}
+          message={msg}
+          header={intl.formatMessage({ id: "actions.clean" })}
+          icon={faTrashAlt}
+          acceptButtonText={intl.formatMessage({ id: "actions.clean" })}
+          acceptButtonVariant="danger"
+          onClose={(p) => {
+            // undefined means cancelled
+            if (p !== undefined) {
+              if (dialogOpen.cleanAlert) {
+                // don't provide paths
+                onClean();
+              } else {
+                onClean(p);
+              }
+            }
+
+            setDialogOpen({
+              clean: false,
+              cleanAlert: false,
+            });
+          }}
+        />
+      );
+    }
+    return;
+  }
```
Why couldn't this have been put into the existing `CleanDialog` instead of removing the `CleanDialog` altogether?

Flashy78 replied:
First I duplicated the `CleanDialog` code over into `LibraryTasks.tsx` as an `AutoTagDialog`.
That was almost identical code, and now a third instance of the `DirectorySelectionDialog` code. So I pulled `CleanDialog` and `AutoTagDialog` out into a new component file.
This new file was almost identical to `DirectorySelectionDialog`, except for the extra styling options, message and option to hide the directory selection.

I thought about adding the styling options to `DirectorySelectionDialog` and replacing the directory code in my new file with that component, but I would have to deconstruct `DirectorySelectionDialog` because it already has a `<ModalComponent>` that I couldn't wrap in my higher level `<ModelComponent>`.

I figured since I was already going to add all the styling options to it, adding message and the option to totally hide it was less complicated and less error prone than trying to refactor `DirectorySelectionDialog` and pull out the directory selection code into a sub component that could be imported into an `OptionalDirectorySelectionDialog` component.

I could do that if you think it would be a better approach?

WithoutPants replied:
> First I duplicated the CleanDialog code over into LibraryTasks.tsx as an AutoTagDialog.
> That was almost identical code, and now a third instance of the DirectorySelectionDialog code. So I pulled CleanDialog and AutoTagDialog out into a new component file.
> This new file was almost identical to DirectorySelectionDialog, except for the extra styling options, message and option to hide the directory selection.

Please do this instead for now. We can address the duplicate code later if needed.

WithoutPants commented on ui/v2.5/src/components/Settings/Tasks/DirectorySelectionDialog.tsx:
```diff
@@ -15,12 +16,29 @@ interface IDirectorySelectionDialogProps {
+  animation,
+  allowEmpty = false,
+  initialPaths = [],
+  allowPathSelection = true,
```
This entire concept is a bit broken from a design perspective. The component is a `DirectorySelectionDialog`, but we have a flag to disable the entire purpose of the dialog? I think there's a better way to do this.

WithoutPants commented on ui/v2.5/src/components/Settings/Tasks/LibraryTasks.tsx:
```diff
@@ -74,17 +74,17 @@ export const LibraryTasks: React.FC = () => {
     useState<GQL.AutoTagMetadataInput>({
-      performers: ["*"],
-      studios: ["*"],
+      performers: [],
+      studios: [],
       tags: ["*"],
```
The reason for specifying tags only as the default should be documented here. Right now, it just looks arbitrary.

WithoutPants commented on ui/v2.5/src/components/Settings/Tasks/LibraryTasks.tsx:
```diff
@@ -210,19 +210,36 @@ export const LibraryTasks: React.FC = () => {
     }
-
-    setDialogOpen({ autoTag: false });
+    return;
```
Redundant

Flashy78 commented:
>     * It would be nice if the Auto Tag dialogs on the studios, performers and tags pages showed different messages, explaining that it will Auto Tag for that specific studio/performer/tag only.
> 
>     * Since there's no big warning icon, I think adding a simple "Are you sure you want to continue?" message to the Auto Tag confirmation dialogs would help to emphasize that the action is irreversible.

Added a line confirming which object you just clicked Auto Tag for.

Took your suggestion and reworked the text into a description and warning, so I could add a warning icon along with the warning text. I've updated the PR description with screenshots.

echo6ix commented:
When I proposed this in https://github.com/stashapp/stash/issues/3909 I never considered a potentially better idea that's already in use.

Instead of a simple warning modal, take inspiration by the deletion modal and provide the user with a list of affected files as a verification layer.

- Autotag is clicked
- Autotag modal window opens
- Modal shows all files that would be affected by the autotag operation
- Apply button commits the autotag operation

```
┌────────────────────────────────────────────────────────┐
│ {$warn_icon}                       Autotag (4 items)   │
├────────────────────────────────────────────────────────┤
│                                                     |s │
│   • path/filename1.jpg                              |c │
│   • path/filename2.jpg                              |r │
│   • path/filename3.jpg                              |o │
│   • path/filename4.jpg                              |l │
│                                                     |l │
│                                                     |b │
│                                                     |a │
│                                                     |r │
│                                                        │
├────────────────────────────────────────────────────────┤
│                [ Apply ]        [ Cancel ]             │
├────────────────────────────────────────────────────────┤
```
This is actually useful now, to new and experienced users who might otherwise just consider a boilerplate modal annoying.

The warning string become redundant now, because it's not some mysterious process happening in the background with unknown consequences. The consequences are enumerated in a reviewable list.

Gykes commented:
I am currently implementing something for this but just keeping it simple. I will add in a warning modal that pops up when auto tag is used to further prevent accidental clicks. With how potentially destructive auto tag is having it just be a click and go bothers me.

WithoutPants commented:
Superceded by #6735
